bftengine/src/bftengine/FullNodeReplica.cpp

```diff
@@ -0,0 +1,295 @@
+    checkpointsInfo.clear();
+    LOG_INFO(GL, "call to startCollectingState()");
+    stateTransfer->startCollectingState();
+  }
+}
+
+void FullNodeReplica::persistCheckpointDescriptor(const SeqNum &seqnum, const CheckpointInfo<false> &chckpinfo) {
+  std::vector<CheckpointMsg *> msgs;
+  msgs.reserve(chckpinfo.getAllCheckpointMsgs().size());
+  for (const auto &m : chckpinfo.getAllCheckpointMsgs()) {
+    msgs.push_back(m.second);
+    LOG_INFO(GL,
+             KVLOG(m.second->seqNumber(),
+                   m.second->epochNumber(),
+                   m.second->state(),
+                   m.second->stateDigest(),
+                   m.second->reservedPagesDigest(),
+                   m.second->rvbDataDigest(),
+                   m.second->idOfGeneratedReplica()));
+  }
+  DescriptorOfLastStableCheckpoint desc(ReplicaConfig::instance().getnumReplicas(), msgs);
+  const size_t bufLen = DescriptorOfLastStableCheckpoint::maxSize(ReplicaConfig::instance().getnumReplicas());
+  concord::serialize::UniquePtrToChar descBuf(new char[bufLen]);
+  char *descBufPtr = descBuf.get();
+  size_t actualSize = 0;
+  desc.serialize(descBufPtr, bufLen, actualSize);
+  ConcordAssertNE(actualSize, 0);
+
+  // TODO [TK] S3KeyGenerator
+  // checkpoints/<BlockId>/<RepId>
+  std::ostringstream oss;
+  oss << "checkpoints/" << msgs[0]->state() << "/" << config_.replicaId;
+  metadataStorage_->atomicWriteArbitraryObject(oss.str(), descBuf.get(), actualSize);
+}
+
+template <>
+void FullNodeReplica::onMessage<ClientRequestMsg>(std::unique_ptr<ClientRequestMsg> msg) {
+  const NodeIdType senderId = msg->senderId();
+  const NodeIdType clientId = msg->clientProxyId();
+  const bool reconfig_flag = (msg->flags() & MsgFlag::RECONFIG_FLAG) != 0;
+  const ReqId reqSeqNum = msg->requestSeqNum();
+  const uint64_t flags = msg->flags();
+
+  SCOPED_MDC_CID(msg->getCid());
+  LOG_DEBUG(CNSUS, KVLOG(clientId, reqSeqNum, senderId) << " flags: " << std::bitset<sizeof(uint64_t) * 8>(flags));
+
+  const auto &span_context = msg->spanContext<std::remove_pointer<ClientRequestMsg>::type>();
+  auto span = concordUtils::startChildSpanFromContext(span_context, "bft_client_request");
+  span.setTag("rid", config_.getreplicaId());
+  span.setTag("cid", msg->getCid());
+  span.setTag("seq_num", reqSeqNum);
+
+  // A full node replica can handle only reconfiguration requests. Those requests are signed by the operator and
+  // the validation is done in the reconfiguration engine. Thus, we don't need to check the client validity as in
+  // the committers
+
+  if (reconfig_flag) {
+    LOG_INFO(GL, "FN replica has received a reconfiguration request");
+    executeReadOnlyRequest(span, *(msg.get()));
+    return;
+  }
+}
+
+void FullNodeReplica::executeReadOnlyRequest(concordUtils::SpanWrapper &parent_span, const ClientRequestMsg &request) {
+  auto span = concordUtils::startChildSpan("bft_execute_read_only_request", parent_span);
```

Comment:
nit: should we add a comment here to check in the future if we are in sync + allowed to execute. If not in-sync, maybe you should reject, do nothing, or ask the validators.

Reply:
This call is called during reconfiguration request, currently fullnode doesn't support reconfiguration, hence i am removing this function and added a TODO statement where this function was being called.